### PR TITLE
fix: Time Column on Generic X-axis

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
Fixes Time Column filters when `GENERIC_CHART_AXES` is enabled. Specifically:
- If a time column is applied and the x-axis is temporal, replace the x-axis and its associated filter with the time column
- If a time column is applied and the x-axis is NOT temporal, replace the first temporal filter with the time column if the time column is not already in the filters
- Moves to the server side the behavior of applying a time range to all temporal filters

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/217339766-c752a47c-7826-4a20-81ef-6ed76426772f.mov

https://user-images.githubusercontent.com/70410625/217339914-a0147934-0caf-4f48-ac15-27be684bd81a.mov

### TESTING INSTRUCTIONS
- Make sure the above rules work as intended
- Make sure different combinations of Time Range and Time column work as intended

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by kgabryje — APPROVED** [review]
LGTM

File: superset/common/query_context_factory.py [hunk 1]
```diff
@@ -99,3 +103,89 @@ def _convert_to_model(self, datasource: DatasourceDict) -> BaseDatasource:
...
+            column.column_name
+            for column in datasource.columns
+            if (column["is_dttm"] if isinstance(column, dict) else column.is_dttm)
+        }
+        granularity = query_object.granularity
+        x_axis = form_data and form_data.get("x_axis")
+
+        if granularity:
+            filter_to_remove = None
+            if x_axis and x_axis in temporal_columns:
```
Comment (zephyring) [inline superset/common/query_context_factory.py:133]: hey @michael-s-molina we recently found a regression where the `x_axis` could be a `dict` while `temporal_columns` is a set of string and it throws.

Steps to reproduce:
```
1. Create a Bar Chart using the Vehicle Sales dataset.
2. Use Custom SQL for the X-Axis field and type in order_date.
3. Use SUM(price_each) for the Metrics field.
4. Use the SQL Lab to duplicate the Vehicle Sales dataset.
5. Perform a dataset swap to the duplicated Vehicle Sales dataset.
6. Update and Save the chart.
```

I am trying to use `x_axis["sqlExpression"]` to get the proper str to compare down the road and it doesn't seem to have the same effect.

Would love your thoughts here thanks!

Comment (michael-s-molina) [reply superset/common/query_context_factory.py:133]: Hi @zephyring. Can you open a draft PR with your changes (`x_axis["sqlExpression"]`)? Can you also add a comment to your PR explaining what do you mean by 

> it doesn't seem to have the same effect

Comment (villebro) [reply superset/common/query_context_factory.py:133]: @michael-s-molina @zephyring It appears the issue here happens due to using an adhoc column as an x-axis. Previously I've also noticed that adhoc x-axes may not always work as expected, and this logic probably needs to be refined somewhat. Maybe we should have a Zoom session to work on a proper fix for this?

Comment (emirkmo) [reply superset/common/query_context_factory.py:133]: Is there an issue or PR linked to this?

**Pull request merged**